### ISSUE TITLE
(V2) I can't fix it, I can't understand why...

### ISSUE THREAD

**Issue by silentlamb1991** [Unity-Technologies/PostProcessing#297]
##### What happened?
I found settings of custom effect always recreated when select the camera, so check the `PostProcessLayerEditor` I found `InitBundles` is always called in OnEnable, so I simply add a condition `if (!m_Target.haveBundlesBeenInited)`, is this right?
No. It causes new issue: `sortedBundles` is null in back code. I check `PostProcessLayer` and found only one place set the `haveBundlesBeenInited` to true, it's in `InitBundles`. I write some debug code to know why.

In `PostProcessLayer`:
```
[UnityEditor.InitializeOnLoadMethod]
static void WhyWhyWhy()
{
	var os = FindObjectsOfType<PostProcessLayer>();
	foreach (var o in os)
	{
		Debug.Log("object " + o.GetInstanceID() + ".haveBundlesBeenInited = " + o.haveBundlesBeenInited);
	}
}
```

In InitBundles method:
```
// Done
haveBundlesBeenInited = true;
Debug.Log("InitBundles");
```

In Editor.OnEnable:
```
Debug.Log("object " + m_Target.GetInstanceID() + ".haveBundlesBeenInited = " + m_Target.haveBundlesBeenInited);

if (!m_Target.haveBundlesBeenInited)
	m_Target.InitBundles();
```

This is output:
![default](https://user-images.githubusercontent.com/18855621/30512187-b539f5da-9b1b-11e7-802c-a8782fb87d17.PNG)

Look, no "InitBundles" output but the bool flag changed!

**Comment by Chman** [comment]
I'm not sure what you're trying to fix exactly. Settings are not recreated, only the bundle stuff is and only in the editor, there's a comment in `PostProcessLayerEditor` explaining why it's needed.

> // In case of domain reload, if the inspector is opened on a disabled PostProcessLayer
> // component it won't go through its OnEnable() and thus will miss bundle initialization
> // so force it there - also for some reason, an editor's OnEnable() can be called before
> // the component's so this will fix that as well.

It's part of a workaround for a serialization / lifetime "issue" in the editor and doesn't affect runtime use at all. You'll find more comments as to why it's done this way in the `PostProcessLayer` class (look for the `SerializedBundleRef` class definition). This is a highly sensible part of the code that is made to work well with the current serialization system in Unity for custom-ordered user effect lists. You **shouldn't** modify it unless you absolutely know what you're doing...

**Comment by Chman** [comment]
See 0cf8662. This is a test fix for the Renderer lifecycle. Might be related.

As written in the comment:
> Expect this to break in some edge cases in the editor... Or not.

**Comment by silentlamb1991** [comment]
I'm trying to write a custom effect, need another command buffer to draw some renderers use replaced material, then use context.command renders post effect.
in Render:
```
if (!_camera)
{
	_camera = context.camera;
	_stencilCommand = new CommandBuffer { name = "Stencil" };
	_camera.AddCommandBuffer(CameraEvent.AfterForwardAlpha, _stencilCommand);
}
// use _stencilCommand draw some renderers
```
In Release:
```
	base.Release();
	if (_camera)
	{
		_camera.RemoveCommandBuffer(CameraEvent.AfterForwardAlpha, _stencilCommand);
		_camera = null;
	}
}
```
Every time I select camera it add a new "Stencil" command.
![ff](https://user-images.githubusercontent.com/18855621/30605422-91a98a7c-9da0-11e7-8105-30d2d9333756.PNG)
I tried to move these code into settings class, same result. So I tried to check haveBundlesBeenInited in OnEnable of editor. The strange thing is, your test fix is working, I can't reproduce the issue I submitted before.
